### PR TITLE
added Floki.attr/4

### PR DESCRIPTION
**Pull request by erikdsi**
I had to change the href's of some links and it seemed easier to be done with Floki's HTMLTree structure so I added a function to do it. I think functions similar to jQuery.text and jQuery.html would be cool too and I may be able to implement that, I'm pretty sure the implementation isn't as clean as it should be, so, thanks for your attention.

### CODE REVIEW

**Comment by philss** [comment]
Hi @erikdsi, thank you for the PR!
I need more time to check this carefully. In the meanwhile, please fix the conflicts listed above. :)

File: lib/floki.ex [hunk 1]
```diff
@@ -153,6 +153,59 @@ defmodule Floki do
     Enum.map(results, fn(html_node) -> HTMLTree.to_tuple(tree, html_node) end)
   end
 
+  defp add_nodes_to_tree(tree, [html_node]) do
```
Comment (philss) [inline lib/floki.ex]: Can you move all the auxiliar functions (private ones) to the bottom of the public `attr/4`?
It's easier to understand what the role of the functions after read the public function that uses them.

File: lib/floki.ex [hunk 2]
```diff
@@ -153,6 +153,59 @@ defmodule Floki do
...
+    Map.put(tree, :nodes, nodes)
+    |> add_nodes_to_tree(tail)
+  end
+
+  defp mutate_attrs(html_tree_list, _, [], _, _) do
+    html_tree_list
+  end
+
+  defp mutate_attrs(_, tree, results, attribute_name, mutation) do
+
```
Comment (philss) [inline lib/floki.ex]: Please remove this empty line.

File: lib/floki.ex [hunk 3]
```diff
@@ -153,6 +153,59 @@ defmodule Floki do
...
+      end
+    end)
+    |> Enum.filter(fn(el) -> el != nil end)
+    |> Enum.map(fn(html_node) -> HTMLTree.to_tuple(tree, html_node) end)
+  end
+  @doc """
+  Changes the attribute values of the elements matched by `selector`
+  with the function `mutation` and returns the whole element tree
+  """
+  def attr(html_str, selector, attribute_name, mutation) when is_binary(html_str) do
```
Comment (philss) [inline lib/floki.ex]: As I said, please move these functions to the top of those private functions.

File: test/floki_test.exs [hunk 4]
```diff
@@ -794,4 +794,25 @@ defmodule FlokiTest do
...
+    <a href=\"http://not.url/changethisbutnotrly/\">link</a>
+    <a class="change" href=\"http://not.url/changed/\">link</a>
+    """
+    result = Floki.attr(html, ".change", "href", fn(inner_html) ->
+      String.replace(inner_html, "changethis", "changed")
+    end)
+    |> Floki.raw_html
+
+    assert result == String.replace(expects, "\n", "")
+
```
Comment (philss) [inline test/floki_test.exs]: you can remove this empty line.

File: test/floki_test.exs [hunk 5]
```diff
@@ -794,4 +794,25 @@ defmodule FlokiTest do
...
+  end
+
```
Comment (philss) [inline test/floki_test.exs]: also, please remove this empty line.

**Comment by erikdsi** [comment]
I applied your suggested fixes and add a typespec and doctest.
Thanks for your attention.

**Comment by philss** [comment]
Hi @erikdsi. I'm about to merge this, but seems that your branch is not updated with master.
Could you rebase with master, please? Thank you.

**Pull request merged**